lib/puppet/environments.rb

```diff
@@ -186,34 +186,37 @@ def self.from_path(path, global_module_path)
       end
     end
 
+    def self.real_path(dir)
+      if Puppet::FileSystem.symlink?(dir) && Puppet[:versioned_environment_dirs]
+        dir = Puppet::FileSystem.expand_path(Puppet::FileSystem.readlink(dir))
+      end
+      return dir
+    end
+
     # @!macro loader_search_paths
     def search_paths
       ["file://#{@environment_dir}"]
     end
 
     # @!macro loader_list
     def list
-      valid_directories.collect do |envdir|
-        name = Puppet::FileSystem.basename_string(envdir).intern
-
+      valid_environment_names.collect do |name|
         create_environment(name)
       end
     end
 
     # @!macro loader_get
     def get(name)
-      if valid_directory?(File.join(@environment_dir, name.to_s))
+      if validated_directory(File.join(@environment_dir, name.to_s))
         create_environment(name)
       end
     end
 
     # @!macro loader_get_conf
     def get_conf(name)
-      envdir = File.join(@environment_dir, name.to_s)
-      if valid_directory?(envdir)
-        return Puppet::Settings::EnvironmentConf.load_from(envdir, @global_module_path)
-      end
-      nil
+      envdir = validated_directory(File.join(@environment_dir, name.to_s))
+      return nil unless envdir
+      return Puppet::Settings::EnvironmentConf.load_from(envdir, @global_module_path)
```

Comment:
fwiw, this could be `if envdir` ... I think?

Reply:
What could? Are you saying I could reverse the order of these `return`s? Or are you saying I should change this to an `if`/`else` like the other one?